salt/utils/etcd_util.py

```diff
@@ -497,11 +532,206 @@ def tree(self, path):
         return ret
 
 
+class EtcdApiV3Adapter(EtcdClient):
+    """
+    Since etcd3 has no concept of directories, this class leaves some methods unimplemented.
+
+    These are: ls and write_directory.
+    """
+
+    ENCODING = "UTF-8"
+
+    def __init__(self, opts, **kwargs):
+        if not HAS_ETCD_V3:
+            raise EtcdLibraryNotInstalled("Don't have etcd3-py, need to install it.")
+        log.debug("etcd_util has the libraries needed for etcd v3")
+
+        super().__init__(opts, **kwargs)
+
+        # etcd3-py uses verify instead of ca_cert
+        self.xargs["verify"] = self.xargs.pop("ca_cert", None)
+        self.client = etcd3.Client(host=self.host, port=self.port, **self.xargs)
+
+    def _decode_from_bytes(self, kv):
+        try:
+            kv.key = (
+                kv.key.decode(encoding=self.ENCODING)
+                if (kv.key and isinstance(kv.key, bytes))
+                else kv.key
+            )
+            kv.value = (
+                kv.value.decode(encoding=self.ENCODING)
+                if (kv.value and isinstance(kv.value, bytes))
+                else kv.value
+            )
+        except AttributeError as err:
+            log.debug("etcd3 decoding error: %s", err)
+        return kv
+
+    def watch(self, key, recurse=False, timeout=0, index=None):
+        ret = {"key": key, "value": None, "changed": False, "mIndex": 0, "dir": False}
+        result = self.read(
+            key, recursive=recurse, wait=True, timeout=timeout, waitIndex=index
+        )
+        if result is not None:
+            result = self._decode_from_bytes(result)
+            ret["key"] = result.key
+            ret["value"] = result.value
+            ret["mIndex"] = getattr(result, "mod_revision", 0)
+            ret["changed"] = True
+        else:
+            return {}
+        return ret
+
+    def get(self, key, recurse=False):
+        if not recurse:
+            result = self.read(key)
+            if isinstance(result, list):
+                return result.pop().value
+            return None
+        return self.tree(key)
+
+    def read(self, key, recursive=False, wait=False, timeout=None, waitIndex=None):
```

Comment:
Strictly because this was the signature beforehand.  I could add another kwarg, but as per my earlier comment with regards to recurse/recursive, I'm not sure what we would want to do here.

Reply:
unless `waitIndex` is the casing that someone would demand/expect from translating direct API calls, I'll go with my previous comment on deprecating old gnarly bits since we're already deprecating things